src/devices/bus/msx/cart/msxdos2.cpp

```diff
@@ -3,49 +3,226 @@
+	save_item(NAME(m_secondary_slot));
+}
+
+image_init_result msx_cart_msxdos2j_device::initialize_cartridge(std::string &message)
+{
+	image_init_result result = msx_cart_msxdos2_base_device::initialize_cartridge(message);
+	if (result != image_init_result::PASS)
+		return result;
+
+	u32 ram_size = cart_ram_region() ? cart_ram_region()->bytes() : 0;
+	if (ram_size != 0 && ram_size != 0x20000 && ram_size != 0x40000)
+	{
+		message = "msx_cart_msxdos2: Region 'ram' has unsupported size.";
+		return image_init_result::FAIL;
+	}
+
+	page(0)->install_view(0x0000, 0x3fff, m_view_page0);
+	page(1)->install_view(0x4000, 0x7fff, m_view_page1);
+	page(2)->install_view(0x8000, 0xbfff, m_view_page2);
+	page(3)->install_view(0xc000, 0xffff, m_view_page3);
+	page(3)->install_readwrite_handler(0xffff, 0xffff, read8smo_delegate(*this, FUNC(msx_cart_msxdos2j_device::secondary_slot_r)), write8smo_delegate(*this, FUNC(msx_cart_msxdos2j_device::secondary_slot_w)));
+
+	// Make sure all views we can select do exist.
+	for (int subslot = 0; subslot < 4; subslot++)
+	{
+		m_view_page0[subslot];
+		m_view_page1[subslot];
+		m_view_page2[subslot];
+		m_view_page3[subslot];
+	}
+
+	// Kanji driver
+	m_view_page1[2].install_rom(0x4000, 0x7fff, cart_rom_region()->base() + 0xc000);
+
+	// MSX-DOS2 rom
+	m_view_page1[3].install_read_bank(0x4000, 0x7fff, m_rombank);
+	m_view_page1[3].install_write_handler(m_bank_address, m_bank_address, write8smo_delegate(*this, FUNC(msx_cart_msxdos2j_device::bank_w)));
+
+	// On-cartridge memory mapper RAM
+	if (ram_size)
+	{
+		m_nr_banks = ram_size / 0x4000;
+		for (int i = 0; i < 4; i++)
+			m_rambank[i]->configure_entries(0, m_nr_banks, cart_ram_region()->base(), 0x4000);
+		m_view_page0[1].install_readwrite_bank(0x0000, 0x3fff, m_rambank[0]);
+		m_view_page1[1].install_readwrite_bank(0x4000, 0x7fff, m_rambank[1]);
+		m_view_page2[1].install_readwrite_bank(0x8000, 0xbfff, m_rambank[2]);
+		m_view_page3[1].install_readwrite_bank(0xc000, 0xffff, m_rambank[3]);
+
+		io_space().install_write_tap(0xfc, 0xfc, "bank0", [this] (offs_t, u8& data, u8){ this->mm_bank_w<0>(data); });
+		io_space().install_write_tap(0xfd, 0xfd, "bank1", [this] (offs_t, u8& data, u8){ this->mm_bank_w<1>(data); });
+		io_space().install_write_tap(0xfe, 0xfe, "bank2", [this] (offs_t, u8& data, u8){ this->mm_bank_w<2>(data); });
+		io_space().install_write_tap(0xff, 0xff, "bank3", [this] (offs_t, u8& data, u8){ this->mm_bank_w<3>(data); });
+	}
+
+	return image_init_result::PASS;
+}
+
+u8 msx_cart_msxdos2j_device::secondary_slot_r()
+{
+	return ~m_secondary_slot;
+}
+
+void msx_cart_msxdos2j_device::secondary_slot_w(u8 data)
+{
+	m_view_page0.select((data >> 0) & 0x03);
+	m_view_page1.select((data >> 2) & 0x03);
+	m_view_page2.select((data >> 4) & 0x03);
+	m_view_page3.select((data >> 6) & 0x03);
+	m_secondary_slot = data;
+}
+
+template <int Bank>
+void msx_cart_msxdos2j_device::mm_bank_w(u8 data)
+{
+	m_rambank[Bank]->set_entry(data % m_nr_banks);
+}
```

Comment:
This won’t work properly if the number of banks isn’t a power of two.  If the number of banks *is* a power of two, you can reduce this to a simple mask rather than a relatively expensive modulo operation.  Please either:
* Ensure the number of banks is a power of two on load and reduce this to a mask rather than a modulo, or
* Use the helpers from bus/generic/slot.h to handle non-power-of-two properly.

The Game Boy cartridges do the latter, e.g. see the `mbc_device_base` class.  The function where it sets up ROM banks is [here](https://github.com/mamedev/mame/blob/mame0251/src/devices/bus/gameboy/cartbase.cpp#L79).

Reply:
Switched to using a mask and configuring the ram banks using map_non_power_of_two.
I already had checks in place to only allow 8 or 16 banks but now the code can handle any number of banks, thanks.